src/main/java/com/instipod/duouniversal/DuoUniversalAuthenticator.java

```diff
@@ -42,17 +42,13 @@ private String getRedirectUrl(AuthenticationFlowContext context, Boolean forceTo
             sessionCode = context.generateAccessCode();
         }
 
-        String baseUrl = context.getHttpRequest().getUri().getBaseUri().toString();
-        baseUrl += "realms/" + URLEncoder.encode(context.getRealm().getName(), StandardCharsets.UTF_8) +
-                "/duo-universal/callback";
-        baseUrl += "?kc_client_id=" +
-                URLEncoder.encode(context.getAuthenticationSession().getClient().getClientId(), StandardCharsets.UTF_8);
-        baseUrl += "&kc_execution=" + URLEncoder.encode(context.getExecution().getId(), StandardCharsets.UTF_8);
-        baseUrl += "&kc_tab_id=" +
-                URLEncoder.encode(context.getAuthenticationSession().getTabId(), StandardCharsets.UTF_8);
-        baseUrl += "&kc_session_code=" + URLEncoder.encode(sessionCode, StandardCharsets.UTF_8);
-
-        return baseUrl;
+        return context.getHttpRequest().getUri().getBaseUri().toString().replaceAll("/+$", "") +
```

Comment:
What is this replaceAll for?

Reply:
This is to ensure that `baseUri` doesn't end with a slash (`/`).
Having multiple consecutive slashes into the URL (e.g. `https://example.org/path//to/somewhere`) can be detected as "malformed URL" by some web servers, making them reply with a `400` code.

Obviously this does not apply to the scheme part.